### ISSUE TITLE
Performance concerns

### ISSUE THREAD

**Issue by arrdem** [weavejester/cljfmt#57]
The clojure core project has 33k LoC in Clojure, of which 25k occurs outside of clojure/core.clj. As hinted in #56, formatting all this takes maybe 6 minutes on my fairly beefy machine, a disproportionate amount of which is spent on clojure/core.clj.

In my testing, running everything but clojure/core.clj (25k lines total) took a mere 1:05s.
Running clojure/core.clj (7.5k lines) alone took 5:51s.

Looking at some yourkit output, it seems like cljfmt is accidentally quadratic on `prior-string` which alone seems to account for a whole lot of wall time. Attached find a plaintext hotspots list and a plaintext list of methods by CPU time. I don't quite understand cljfmt yet, but it seems like `prior-string` is being repeatedly re-evaluated on forms which have already been fully reindented.

[Method-list-by-own-CPU-time.txt](https://github.com/weavejester/cljfmt/files/144147/Method-list-by-own-CPU-time.txt)
[CPU-hot-spots.txt](https://github.com/weavejester/cljfmt/files/144148/CPU-hot-spots.txt)

**Comment by weavejester** [comment]
When I was writing cljfmt, I quickly found that writing a Clojure code formatter, even with a tool like rewrite-clj, was a lot more complex than I had anticipated. For this reason I prioritised clearer, simpler code over performance, based on the idea that I could always come back later. For most Clojure programs, cljfmt turned out to be fast enough, so I haven't revisited performance.

The `prior-string` function in particular is very inefficient. It's designed purely to find the column, or "margin" of a zipper location. Although rewrite-clj includes column metadata of the original source, that metadata doesn't automatically change, and quickly becomes outdated. The `prior-string` function works it's way up and left the zipper, constructing the output string prior to the position of the zipper.

For example, assume we have some input:

```
(let [x 1
      y 2]
  (+ x 1)
```

And let's say the zipper is sitting at `y`.

In which case, the prior string is:

```
"(let [x 1\n      "
```

We get the last line in the string, count it, and come out with an answer of 6. That's the column number of `y`.

One way of getting around this inefficiency is to keep track of the column with each change in the zipper. So if I change a node, I'd also work out the effect it has on other nodes and update the column numbers appropriately. However, this seems somewhat complex and prone to error.

An rather easier option is simply to memoize the `prior-string` function. The majority of the time it's being called with calculations it's already done previously. My expectation is that memoizing the function will have a significant performance improvement at the cost of very little work.

TLDR; try memoizing `prior-string` and see if that helps.

**Comment by weavejester** [comment]
Should be fixed in 0.4.1

**Comment by arrdem** [comment]
Given that cljfmt is still about 20x slower than Java format linters I think there's still plenty to be done, but the obvious pain point has certainly been taken care of. I'll open new tickets/PRs if/when I have time to work on this more.

**Comment by weavejester** [comment]
Sure. I mostly just didn't want an open-ended performance issue, as that would never be resolved :)
